### PR TITLE
Constraint stream docs: grouping and collectors

### PR DESCRIPTION
**Pull request by triceo**

### CODE REVIEW

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 1]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
                         Talk::getDurationInMinutes);
     }
 ----
+
+[[constraintStreamsGroupingAndCollectors]]
+== Grouping and Collectors
+
+In the constraint streams framework, grouping is a mechanism to group items in a stream according to user-provider
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: "In the constraint streams framework" -> it's a framework now? :)
I'd remove that context, as it's inside the ConstraintStreams session.

"grouping is a mechanism to group items in a stream according to ..."
Docs style suggestions: keep it short and use active voice, so instead:
"grouping collects items in a stream according to ..."
https://developers.google.com/style/voice

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 2]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+criteria, similar to what a `GROUP BY` SQL clause does. Additionally, some grouping operations also accept one or more
+`Collector`s, which provide various aggregation functions.
+
+For example, the following code snippet will first group all processes by the computer they run on, sum up all the
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: remove "will" (future tense)
https://developers.google.com/style/voice

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 3]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+power required by the processes on that computer using the `ConstraintCollectors.sum(...)` collector, and finally
+penalize every computer whose processes consume more power than is available.
+
+[source,java,options="nowrap"]
+----
+    private Constraint requiredCpuPowerTotal(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(CloudProcess.class)
+                .groupBy(CloudProcess::getComputer, sum(CloudProcess::getRequiredCpuPower))
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: The filter(reqCpu > computer.getCpu()) is missing,
so the description does not match the implementation "whose processes consume more power than is available".

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 4]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+                .penalize("requiredCpuPowerTotal",
+                        HardSoftScore.ONE_HARD,
+                        (computer, requiredCpuPower) -> requiredCpuPower - computer.getCpuPower());
+    }
+----
+
+OptaPlanner provides several collector implementations you can use in your grouping constraint streams. This chapter
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: "OptaPlanner provides" -> there are
"you can use" -> "to"
"in your grouping constraint streams" -> singular is more accurate. Per one, you can still use multiple collectors.
->
"There are several collectors out-of-the-box when grouping a constraint stream."

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 5]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+discusses the most important ones. For a complete list, refer to the
+`org.optaplanner.core.api.score.stream.ConstraintCollectors` class.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: This section (not chapter) must describe all of them, so we can leave this sentence out.
Do note that this section will only describe "sum(int)" in detail, and simply mention there are "variants for long, BigDecimal and Duration" (without describing those in detail).

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 6]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+        return factory.from(Talk.class)
+                .groupBy(Talk::hasAnyUnavailableSpeaker, count())
+                .penalize("speakerAvailability",
+                        HardSoftScore.ONE_HARD,
+                        (hasUnavailableSpeaker, count) -> ...);
+    }
+----
+
+The `ConstraintCollectors.countDistinct(...)` will do the same, but it will take object equality into account. Different
+instances of objects which still `equals(...)` will therefore only count as one instance.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: code example of countDistinct please
I'd argue that countDistinct needs it owns section

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 7]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+
+OptaPlanner provides several collector implementations you can use in your grouping constraint streams. This chapter
+discusses the most important ones. For a complete list, refer to the
+`org.optaplanner.core.api.score.stream.ConstraintCollectors` class.
+
+You can provide your own collectors by implementing the
+`org.optaplanner.core.api.score.stream.uni.UniConstraintCollector` interface or their `Bi...`, `Tri...` counterparts.
+The following sections will focus on the collector implementations provided out of the box.
+
+=== Counting collectors
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: code style for product-docs friendliness: All Words in a Title Must Start with a Capital unless They Are In-between-words.
=> "Counting Collectors"

Counting != count()
I'd avoid the proliferation of additional ways of saying the same thing, it creates confusion.
Can we rename these sections to:
- Collecting `count()`
- Collecting `countDistinct()`
- Collecting `sum()`
etc?
Or just
- `count()`
- `countDistinct()`
- `sum()`

Comment (ge0ffrey) [reply optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: Every title must have a unique ID, usually camel case of the title, sometimes prefixed with the parent section's title

[collectorsCount]
=== `count`


[countingCollectors]
=== Counting collectors

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 8]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
                         Talk::getDurationInMinutes);
     }
 ----
+
+[[constraintStreamsGroupingAndCollectors]]
+== Grouping and Collectors
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: in this section, I'd make 2 subsections, besides the introduction with a first example:
- group key's
- collectors

That way, all the build-in collectors, are grouping by a single section, separate from the group key one

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 9]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+                        HardSoftScore.ONE_HARD,
+                        (hasUnavailableSpeaker, count) -> ...);
+    }
+----
+
+The `ConstraintCollectors.countDistinct(...)` will do the same, but it will take object equality into account. Different
+instances of objects which still `equals(...)` will therefore only count as one instance.
+
+The return value for these collectors is a 32-bit signed integer (`int`). Should that not be enough for you, please
+refer to the `countLong()` and `countDistinctLong()` variants of these collectors.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: Can we describe the variants consistently similar to how we do it in here?
  https://github.com/kiegroup/optaplanner/blob/master/optaplanner-docs/src/main/asciidoc/ScoreCalculation/ScoreCalculation-chapter.adoc#hardSoftScore
Variants include: uni, bi, tri.
Variants include: int, long, bigdecimal, duration.

(I don't consider countDistinct() a variant).

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 10]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+
+=== Summing collector
+
+Summing collectors will sum the value of a particular property of all elements in the group. For an example of using a
+`ConstraintCollectors.sum(...)` collector, <<constraintStreamsGroupingAndCollectors,see above>>.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: show a code example here too, but don't bother adding the method name etc.
The main purpose of the code example is to:
- 1) complete the "fit for purpose" check in the eyes of the read
- 2) copy-paste it if it fits for purpose.

1) By looking at sum() and count() examples, the reader can understand the difference by example, instead by abstraction. Some people learn by example, others learn by abstract (bottom up vs top down). So we need to cater for both. Once the reader understands the difference, he/she can make a "fit for purpose" decision.

2) And it's oh so convenient if he can just copy-paste it if it is indeed fit for purpose :)

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 11]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+
+The return value for this collector is a 32-bit signed integer (`int`). Should that not be enough for you, please
+refer to the `sumLong(...)` variant of this collector.
+
+=== Minimums and maximums
+
+In certain situations, it may be necessary to find a minimal or a maximal element in a group. For that, we provide
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: remove "in certain situations". it implies that this is pretty rare or special, which I don't think "min()" nor "max" is.

remove "we provide" -> active voice. 

-> "To extract the minimum or maximum of a group, use the `ConstraintCollectors.min(...)` and `ConstraintCollectors.max(...)` collectors respectively."

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 12]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+`ConstraintCollectors.min(...)` and `ConstraintCollectors.max(...)` collectors respectively.
+
+These collectors operate on values of properties which are `Comparable` (such as `Integer`, `String` or `Duration`),
+although we also provide variants of these collectors which allow you to provide your own `Comparator`.
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: "we" -> active voice

note: in blog posts, the use of "we" is ok. In docs, we don't use it normally. The use of "you" is avoided, but there are many cases where it is needed, and can't be avoided decently with a pure active voice.

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc [hunk 13]
```diff
@@ -57,3 +57,86 @@ and supports <<explainingTheScore,justifications>>:
...
+
+Here's an example of finding a computer which runs the most power-demanding process:
```
Comment (ge0ffrey) [inline optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: "Here's an example" -> "For example," (short & sweet)

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc (hunk 9 above)
Comment (triceo) [reply optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: I don't think we should. The variants are easy to add, and we will be adding more as we see fit. People should refer to the Java API for a complete list - duplicating this in the docs is a recipe for pain and, more importantly, obsoleteness.

**Comment by triceo** [comment]
@ge0ffrey Updates made.

File: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc (hunk 9 above)
Comment (ge0ffrey) [reply optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc]: I fully agree. We can add a short section in the beginning that everything comes in Uni/Bi/Tri etc variants later, if needed.

**Pull request merged**